cpp/src/arrow/util/decimal.cc

```diff
@@ -42,34 +45,40 @@ namespace arrow {
 using internal::SafeLeftShift;
 using internal::SafeSignedAdd;
 
+#ifdef __SIZEOF_INT128__
+using uint128_t = __uint128_t;
+#else
+using boost::multiprecision::uint128_t;
+#endif
```

Comment:
Can we centralize this definition in e.g. `src/arrow/util/int128_internal.h`?

Reply:
I'll note that `src/arrow/util/decimal_test.cc` also uses `int128_t`.

Reply:
Moved to int128_internal.h. Unfortunately, __int128_t and boost int128_t are not interchangeable. For example, operator<<(std::ostream, __int128_t) is not defined, which means I can't replace int128_t in decimal_test.cc. In addition, there are some subtle behavior differences or perhaps bugs in boost int128_t (e.g., static_cast<uint64_t>(boost int128_t) returns the max value of uint64_t when the input has more than 64 bits in some of the configs, while static_cast<uint64_t>(__int128_t) returns the lower 64 bits). Do you think it's better to leave this alias within decimal.cc due to the limitations?

Also reverted the changes to testing submodule.

Reply:
Hmm, surely we can avoid using `operator<<`? We can just note the limitations in the header file. It would be internal anyway.

Reply:
Updated the test and added comments. Also fixed some typos in the tests that test the wrong operators.